Fixed bug that some Multiblock machines unnecessarily recheck their structure when they get redstone updates via `onNeighborChange` if Alternate Current is installed.

MineCake147E opened the pull request:
## Changes proposed in this pull request:

<img width="3840" height="2036" alt="Redstone Wire on top of the Reactor Glass" src="https://github.com/user-attachments/assets/ab41d458-5027-430d-8515-3207f83278e0" />

When the Redstone Wire on top of the Reactor Glass gets updated, `TileEntityInternalMultiblock` beneath the `mekanismgenerators:reactor_glass` (in this case the `TileEntityControlRodAssembly`) receives `onNeighborChange(Block block, BlockPos neighborPos)` with `block` set to `Block{minecraft:redstone_wire}`.  
If [Alternate Current](https://modrinth.com/mod/alternate-current) is not present, `neighborPos` will be the position of the `mekanismgenerators:reactor_glass` in this case.  
However, if it is present, `neighborPos` will be the position of the `minecraft:redstone_wire`, which is right above the `mekanismgenerators:reactor_glass` in this case.  
This causes the `multiblock.isKnownLocation(neighborPos)` return `false`, thus setting `multiblock.recheckStructure` to `true`, ultimately triggering the multiblock structure recheck.  

<img width="3840" height="2036" alt="Redstone Wire connected to the Fission Reactor Logic Adapter on top of the Reactor Glass" src="https://github.com/user-attachments/assets/7b917a42-e80b-4939-839e-b136c9bd1923" />
<img width="3840" height="2036" alt="Configuration of Fission Reactor Logic Adapter (Insufficient Fuel)" src="https://github.com/user-attachments/assets/2256dacc-b126-47b6-94c0-f00e265e3074" />

This setup above makes it even worse. The Redstone mode of Fission Reactor Logic Adapter is set to Insufficient Fuel here, and the reactor has no fuel this time. Redstone Wire connects the Logic Adapter and the Reactor Glass on top of the Control Rod Assembly.  
This setup causes infinite multiblock structure recheck like in the video below if Alternate Current is present.

https://github.com/user-attachments/assets/9435b5ba-fb16-4bd8-b4c5-7a30ca8997b2

This fix checks if the touching block can be a part of multiblock structure at all, reducing unnecessary multiblock structure rechecks caused by redstone updates even if the Alternate Current is present.

MineCake147E committed:
Fixed bug that Multiblocks unnecessarily recheck their structure when they get redstone updates via `onNeighborChange` if Alternate Current is installed.

thiakil reviewed (changes requested):
> However, if it is present, neighborPos will be the position of the minecraft:redstone_wire, which is right above the mekanismgenerators:reactor_glass in this case.

The redstone wire is not a neighbour of the block BELOW the glass. This is a bug in Alternate Current.

A fix I would accept would be to return early if the `neighborPos` is not `pos +- 1`

MineCake147E committed:
`TileEntityInternalMultiblock.onNeighborChange` no longer checks the type of the `TileEntity` at `neighborPos`.

MineCake147E commented:
> A fix I would accept would be to return early if the `neighborPos` is not `pos +- 1`

I updated the logic accordingly.
EDIT: I also removed the `long` cast that I once thought necessary because [`Math.Abs(int)` in C#](https://learn.microsoft.com/en-us/dotnet/api/system.math.abs?view=net-8.0#system-math-abs(system-int32)) throws when it overflows, but today I learned that Java doesn't do that. Sorry for the inconvenience.

MineCake147E committed:
`TileEntityInternalMultiblock.onNeighborChange` no longer casts the position difference to `long`.

MineCake147E committed:
Fixed bug that `TileEntityInternalMultiblock.onNeighborChange` may cause multiblock structure recheck when the difference of one or more of X, Y, or Z coordinates are `Integer.MIN_VALUE` and others are 1.

thiakil commented on src/main/java/mekanism/common/tile/prefab/TileEntityInternalMultiblock.java:
```diff
@@ -63,16 +63,21 @@ public void onNeighborChange(Block block, BlockPos neighborPos) {
+        int dX = Math.abs(neighborPos.getX() - worldPosition.getX());
+        int dY = Math.abs(neighborPos.getY() - worldPosition.getY());
+        int dZ = Math.abs(neighborPos.getZ() - worldPosition.getZ());
+        if (((dX | dY | dZ) >>> 1) > 0) return;
```
while a neat maths trick, it's far less readable than checking the individual components

thiakil commented on src/main/java/mekanism/common/tile/prefab/TileEntityInternalMultiblock.java:
```diff
@@ -63,16 +63,21 @@ public void onNeighborChange(Block block, BlockPos neighborPos) {
+        int dX = Math.abs(neighborPos.getX() - worldPosition.getX());
+        int dY = Math.abs(neighborPos.getY() - worldPosition.getY());
+        int dZ = Math.abs(neighborPos.getZ() - worldPosition.getZ());
+        if (((dX | dY | dZ) >>> 1) > 0) return;
```
while a neat maths trick, it's far less readable than checking the individual components

thiakil replied:
nb: `net.minecraft.core.Vec3i#distManhattan` would be useful

MineCake147E replied:
I updated the code accordingly.

thiakil commented on src/main/java/mekanism/common/tile/prefab/TileEntityInternalMultiblock.java:
```diff
@@ -63,16 +63,21 @@ public void onNeighborChange(Block block, BlockPos neighborPos) {
+        if (level == null || multiblock == null || isRemote()) return;
+        //If the neighbor change happened to a block inside a multiblock, and it isn't a block that is part of the multiblock
+        if (!level.isEmptyBlock(neighborPos) && multiblock.isKnownLocation(neighborPos)) return;
+        //If the neighbor change happened to a block outside a multiblock, and it isn't a block that could be part of the multiblock
```
Either return the nested `if` statements, or rewrite the comments to make sense (all three sections)

MineCake147E replied:
I updated the code accordingly.

MineCake147E committed:
* `TileEntityInternalMultiblock.onNeighborChange` no longer uses bitwise math tricks to validate the `neighborPos`.
* `TileEntityInternalMultiblock.onNeighborChange` now uses nested `if` statements again.

MineCake147E committed:
* `TileEntityInternalMultiblock.onNeighborChange` now uses Manhattan distance instead of Chebyshev distance.
